src/visitors/sympy_replace_solutions_visitor.hpp

```diff
@@ -0,0 +1,465 @@
+
+        /// Number of interleaves. We need to remove the first activation of the switch
+        inline size_t n() const {
+            return n_ == 0 ? 0 : n_ - 1;
+        }
+
+      private:
+        /**
+         * \brief Number of interleaves of assignment statements in between equations of the system
+         * of equations
+         *
+         * This is equivalent to the number of switches false -> true of \ref in_system_ minus the
+         * very first one (if the system exists).
+         */
+        size_t n_ = 0;
+
+        /// Bool that keeps track if just wrote an equation of the system of equations (true) or not
+        /// (false)
+        bool in_system_ = false;
+    };
+
+
+    /**
+     * \struct StatementDispenser
+     * \brief Sorts and maps statements to variables keeping track of what needs updating
+     *
+     * This is a multi-purpose object that:
+     *
+     * - keeps track of what was already updated
+     * - decides what statements need updating in case there was a variable assignment (i.e. \f a =
+     * 3 \f)
+     * - builds the statements from a vector of strings
+     *
+     */
+    struct StatementDispenser {
+        /// Empty ctor
+        StatementDispenser() = default;
+
+        /// Standard ctor
+        StatementDispenser(const std::vector<std::string>::const_iterator& statements_str_beg,
+                       const std::vector<std::string>::const_iterator& statements_str_end,
+                       const int error_on_n_flushes);
+
+        /**
+         * /brief Construct the maps for easy access and classification of the statements
+         *
+         * Here we construct a map variable -> affected equations. In other words this map tells me
+         * what equations need to be updated when I change a particular variable. To do that we
+         * build a a graph of dependencies var -> vars and in the mean time we reduce it to the root
+         * variables. This is ensured by the fact that the tmp variables are sorted so that the next
+         * tmp variable may depend on the previous one. Since it is a relation of equivalence (if an
+         * equation depends on a variable, it needs to be updated if the variable changes), we build
+         * the two maps at the same time.
+         *
+         * An example:
+         *
+         *  - \f tmp0 = x + a \f
+         *  - \f tmp1 = tmp0 + b \f
+         *  - \f tmp2 = y \f
+         *
+         * dependency_map_ should be (the order of the equation is unimportant since we are building
+         * a map):
+         *
+         * - tmp0 : x, a
+         * - tmp1 : x, a, b
+         * - tmp2 : y
+         *
+         * and the var2statement_ map should be (the order of the following equations is unimportant
+         * since we are building a map. The number represents the index of the original equations):
+         *
+         * - x : 0, 1
+         * - y : 2
+         * - a : 0, 1
+         * - b : 1
+         */
```

Comment:
nitpick : In other parts of the code, we are keeping `/brief` in the header file but the `/details` in the implementation .cpp file. This way, you have logic near the implementation and it's easy to refer / scroll to an update?